app/SpiceHarvesterHarvest.php

```diff
@@ -63,5 +65,51 @@ public function getStatusClassAttribute()
         return 'default';
     }
 
+    public function process(callable $onProcess)
+    {
+        $progress = new Progress();
+
+        try {
+            $this->status = SpiceHarvesterHarvest::STATUS_IN_PROGRESS;
+            $this->status_messages = trans('harvest.status_messages.started');
+            $this->save();
+
+            $onProcess($progress);
+
+            $this->status = SpiceHarvesterHarvest::STATUS_COMPLETED;
+            $this->completed = date('Y-m-d H:i:s');
+            $this->status_messages = trans('harvest.status_messages.completed', [
+                'processed' => $progress->getProcessed(),
+                'created' => $progress->getInserted(),
+                'updated' => $progress->getUpdated(),
+                'deleted' => $progress->getDeleted(),
+                'skipped' => $progress->getSkipped(),
+                'failed' => $progress->getFailed(),
+                'time' => Carbon::now()->diffInSeconds($progress->getCreatedAt()),
+            ]);
+        } catch (\Exception $e) {
+            $this->status = SpiceHarvesterHarvest::STATUS_ERROR;
+            $this->status_messages = trans('harvest.status_messages.error', [
+                'error' => $e->getMessage(),
+            ]);
+        } finally {
+            $this->save();
+        }
+    }
+
+    public function advance(Progress $progress)
```

Comment:
This indicates a 'move forward' action (when in fact, it just reflects the state of `$progress`)
```suggestion
    public function updateStatusMessages(Progress $progress)
```